### DISCUSSION TITLE
Side effects inside a factory capsule?

### DISCUSSION THREAD

**Discussion by opsb** [GregoryConrad/rearch-dart#205] [Q&A] [answered]
I have the following

```dart
AsyncValue<File> Function(String uri) imageStoreImage(CapsuleHandle use) {
  final env = use(syncEnvCapsule);

  return (uri) {
    return use.future(env.imageStore.openImage(uri));
  };
}
```

which I intend to use as 

```dart
Widget _image(BuildContext context, WidgetHandle use) {
    final asyncImage = use(imageStoreImage)(uri);

    return asyncImage
        .map<Widget>((data) => Image.file(data, width: 140, height: 140, fit: BoxFit.cover))
        .dataOr(Text("Loading ..."));
  }
```

The issue is that the Rearch wants the `use.future` call to be moved into the outer `imageStoreImage` function. I considered creating a side effect instead but then there's a dependency on a capsule so this doesn't seem right either. How should I approach this scenario?

**Comment by GregoryConrad** [comment, marked as answer]
Easy fix! Just move the side effect to whatever consumes the factory/action capsule.

```dart
Future<File> Function(String uri) imageStoreImage(CapsuleHandle use) {
  final env = use(syncEnvCapsule);
  return env.imageStore.openImage;
}

Widget _image(BuildContext context, WidgetHandle use) {
    // See comment below for why this wouldn't work directly:
    // final imageFuture = use(imageStoreImage)(uri);

    return use.future(imageFuture)
        .map<Widget>((data) => Image.file(data, width: 140, height: 140, fit: BoxFit.cover))
        .dataOr(Text("Loading ..."));
}
```

**Reply by GregoryConrad** [reply]
Oops, should also mention that you should cache the future (based on when `uri` changes. Create the future like so:

```dart
final imageFuture = use.memo(() => use(imageStoreImage)(uri), [uri]);
```

That's because whenever `use.future` gets a new future (which would happen if you call the function that gives you a future every build), it resets the state and starts watching that new future. So you need to cache the future until it actually needs to change.
